Add <Form> component for easier mutations (#1552)
Shopify/hydrogen: docs/framework/forms.md

---
gid: 9120943b-01c9-4da3-a201-5a54cab6ca2a
title: Forms
description: Declarative mutations with `<Form>` and API Routes
---

> Within a Hydrogen App, [Server components](https://shopify.dev/custom-storefronts/hydrogen/framework/work-with-rsc#fetching-data-on-the-server) are used to fetch data and [API Routes](https://shopify.dev/custom-storefronts/hydrogen/framework/routes#api-routes) to mutate data. The `<Form>` component provides a declarative way to send data to API Routes, and re-render server components.

Reviewer: ```suggestion
Within a Hydrogen App, [Server components](https://shopify.dev/custom-storefronts/hydrogen/framework/work-with-rsc#fetching-data-on-the-server) are used to fetch data and [API Routes](https://shopify.dev/custom-storefronts/hydrogen/framework/routes#api-routes) to mutate data. The `Form` component provides a declarative way to send data to API Routes, and re-render server components.
```

Looking at the Hydrogen component docs, I don't see delimiters around the component names. If this edit is valid, pls apply throughout the PR.

}
```

{% endcodeblock %}

## `<Form>` requires an API Route

In order to properly use the `<Form>` component, the `action` attribute must point to an API route. An example implementation might look like:

{% codeblock file, filename: 'login.server.jsx' %}

```tsx
export async function api(request, {session}) {
  // Access the form data
  const data = await request.formData();
  const username = data.get('username');
  const password = data.get('password');

  // Find the user
  const userId = await getUser(username, password);

  if (!userId) {
    // We couldn't find the user.
    // Re-render the login page with a login error displayed
    return new Request('/login?error');
  }

  // Save the user to the session
  await session.set('userId', userId);

  // Forward the user to the account page
  return new Request('/account');
}
```

{% endcodeblock %}

Read data in the API route from the `<Form>` by using the [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) API. The API Route must respond with a `new Request()`. This renders the server components for the given page. You can re-render the current page, or render an entirely different page in the app.

> In the above example, when the user is not found, we re-render the current page with a search parameter. Let's update the server component to render the login error:

Reviewer: ```suggestion
In the previous example, when the user is not found, the current page is re-rendered with a search parameter. The following code updates the server component to render the login error:
```

      </label>
      {url.searchParams.get('error') ? (
        <h2 className="text-red-700">Invalid username or password</h2>
      ) : null}
      <button type="submit">Submit</button>
    </Form>
  );
}
```

{% endcodeblock %}

Because the `<Form>` is now within a client component, `children` can be a render prop, which allows us to give feedback while the form is submitting:

{% codeblock file, filename: 'LoginForm.client.jsx' %}

```tsx
import {Form} from '@shopify/hydrogen';

export default function Login() {
  const url = useUrl();
  return (
    <Form action="/login" method="post">
      {({loading}) => (
        <>
          <label>
            Username <input type="text" name="username" />
          </label>
          <label>
            Password <input type="password" name="password" />
          </label>
          {url.searchParams.get('error') ? (
            <h2 className="text-red-700">Invalid username or password</h2>
          ) : null}
          <button type="submit" disabled={loading}>
            Submit
          </button>
        </>
      )}
    </Form>
  );
}
```

{% endcodeblock %}

## Hidden fields

> The `<Form>` component can be used for any mutation, not just ones with a text field. For example, maybe when a customer adds an item to a cart, or when they log out:

Reviewer: ```suggestion
You can use the `Form` component for any mutation that doesn't include a text field. 

For example, the following uses a `Form` component for adding items an item to a cart:
```

{% codeblock file, filename: 'Product.server.jsx' %}

```tsx
import {Form} from '@shopify/hydrogen';

export default function Product({product}) {
  return (
    <ProductDetails>
      <Form action="/addToCart" method="post">
        <input type="hidden" name="productId" value={product.id} />
        <button type="submit">Add to cart</button>
      </Form>
    </ProductDetails>
  );
}
```

{% endcodeblock %}

> The hidden input field for the `productId` will be sent to the server when the add to cart button is pressed. The API route at `/addToCart` can contain all the logic to add the product to the cart, then re-render the page. The advantage here is the button will be actionable before the page has fully loaded and JavaScript hydrated.

Reviewer: ```suggestion
The hidden input field for the `productId` is sent to the server when the **Add to cart** button is clicked. The API route at `/addToCart` can contain all the logic to add the product to the cart and re-render the page. The button is actionable before the page fully loads and the JavaScript is hydrated.
```
